In [1]:
import hashlib
import json
from time import time

class Blockchain(object):
    def __init__(self):
        self.chain = []
        self.pending_transactions = []
        
        self.new_block(previous_hash="The Times 03/Jan/2009 Chancellor on brink of second bailout for banks.", proof=100)
        
    def new_block(self,proof,previous_hash=None):
        block = {
            'index': len(self.chain) + 1,
            'timestamp': time(),
            'transactions': self.pending_transactions,
            'proof': proof,
            'previous_hash': previous_hash or self.hash(self.chain[-1]),
        }
        self.pending_transactions = []
        self.chain.append(block)
        
        return block
    
    @property
    def last_block(self):
        
        return self.chain[-1]
    
    def new_transaction(self, sender, recipient, amount):
        transaction = {
            'sender': sender,
            'recipient': recipient,
            'amount': amount
        }
        self.pending_transactions.append(transaction)
        return self.last_block['index'] + 1
    
    def hash(self, block):
        string_object = json.dumps(block, sort_keys=True)
        block_string = string_object.encode()
        
        raw_hash = hashlib.sha256(block_string)
        hex_hash = raw_hash.hexdigest()
        
        return hex_hash

Raw basics to perpetuate the chain:
1. new_transaction() - Allows the token (i.e. Bitcoin) to be sent between addresses    
2. new_block() - Groups all transactions since the last block into a new block    
3. hash() - This is the magic that makes everything 'uncrackable' and verifiable!

In [2]:
blockchain = Blockchain()
t1 = blockchain.new_transaction("Satoshi","Bert",'2 BTC')
t2 = blockchain.new_transaction("Bert","Dr. Cook",'1 BTC')
t3 = blockchain.new_transaction("Bert","Bernie",'0.5 BTC')
blockchain.new_block(12345)

t4 = blockchain.new_transaction("Mike","Alice",'0.25 BTC')
t5 = blockchain.new_transaction("Alice","Bob",'0.1 BTC')
t6 = blockchain.new_transaction("Bob","Mike",'0.1 BTC')
blockchain.new_block(6789)

print("Blockchain: ", blockchain.chain)

Blockchain:  [{'index': 1, 'timestamp': 1617213161.3660836, 'transactions': [], 'proof': 100, 'previous_hash': 'The Times 03/Jan/2009 Chancellor on brink of second bailout for banks.'}, {'index': 2, 'timestamp': 1617213161.3662148, 'transactions': [{'sender': 'Satoshi', 'recipient': 'Bert', 'amount': '2 BTC'}, {'sender': 'Bert', 'recipient': 'Dr. Cook', 'amount': '1 BTC'}, {'sender': 'Bert', 'recipient': 'Bernie', 'amount': '0.5 BTC'}], 'proof': 12345, 'previous_hash': '5619d74469fd77c2c282cf785bea4131c7e253864ec206df78ba621e137e88b6'}, {'index': 3, 'timestamp': 1617213161.366364, 'transactions': [{'sender': 'Mike', 'recipient': 'Alice', 'amount': '0.25 BTC'}, {'sender': 'Alice', 'recipient': 'Bob', 'amount': '0.1 BTC'}, {'sender': 'Bob', 'recipient': 'Mike', 'amount': '0.1 BTC'}], 'proof': 6789, 'previous_hash': '543da9963d3779f9a47548644bbfbb9a0f21244fadbb6a8c338518c3532381b7'}]


For better visibility:

Blockchain:  [

    {'index': 1, 
    'timestamp': 1617213161.3660836, 
    'transactions': [], 
    'proof': 100, 
    'previous_hash': 'The Times 03/Jan/2009 Chancellor on brink of second bailout for banks.'}
    , 
    
    {'index': 2, 
    'timestamp': 1617213161.3662148, 
    'transactions': [
        {'sender': 'Satoshi', 'recipient': 'Bert', 'amount': '2 BTC'}, 
        {'sender': 'Bert', 'recipient': 'Dr. Cook', 'amount': '1 BTC'}, 
        {'sender': 'Bert', 'recipient': 'Bernie', 'amount': '0.5 BTC'}], 
    'proof': 12345, 
    'previous_hash': '5619d74469fd77c2c282cf785bea4131c7e253864ec206df78ba621e137e88b6'}, 
    
    {'index': 3, 
    'timestamp': 1617213161.366364, 
    'transactions': [
        {'sender': 'Mike', 'recipient': 'Alice', 'amount': '0.25 BTC'}, 
        {'sender': 'Alice', 'recipient': 'Bob', 'amount': '0.1 BTC'}, 
        {'sender': 'Bob', 'recipient': 'Mike', 'amount': '0.1 BTC'}], 
    'proof': 6789, 
    'previous_hash': '543da9963d3779f9a47548644bbfbb9a0f21244fadbb6a8c338518c3532381b7'}
    ]

The set of pending transactions are hashed and added to the hash of the previous block. At any point in time the ledger can be 'deconstructed' to show all transactions of previous blocks all the way back to the genesis block.

![](https://miro.medium.com/max/4692/1*_RY4v9D5-BOVMzFMr8Dk4A.png)

One piece missing from the above graphic (a very important piece for the Bitcoin blockchain) is the **nonce** - a unique value that will allow the hashed output to have a set number of zeros. This number of zeros can be adjusted up or down by the network itself to make the transaction more or less difficult (in attempt to keep computation time close to constant: ~10 minutes for Bitcoin).

![](https://www.tutorialspoint.com/assets/questions/media/22879/bitcoin_block_hashing.jpg)

Every miner in the network is racing to find the right nonce to add to their pending transactions. Once one miner cracks this nonce puzzle and a set number of other miners verify this new hash block it gets published to the blockchain and the process starts anew!

This process is called **Proof of Work** (PoW). The new hash block is proof that the miner went through the laborious steps of incorporating previous chain data, adding new transactions, and making it all fit the pattern. As a reward for this work the miner receives some of the token (i.e. Bitcoin) as a reward.

Advantages of this network structure:
* Byzantine Fault Tolerance
* Immutable ledger
* Transparent(-ish)
* "Trustless"
* **Decentralized**

Disadvantages of this network structure:
* Energy-inefficient
* Costly to scale
* Redundant work
* Vulnerable to 51% attacks


Due to these disadvantages a lot of work has gone into improving this network structure with new consensus algorithms.
Variations to PoW:
* Proof of Stake (PoS)
* Proof of Authority (PoA)
* Delegated Proof of Stake (DPoS)

Other applications of blockchain technology:
* Supply Chain Ledgers
    * OriginTrail
    * VeChain
* Medical Record Management
    * MedicalChain
    * Patientory
* Distributed AI
    * Numeraire
    * Ocean Protocol
    * CureCoin
* Non-Fungible Tokens
    * Decentraland
    * Enjin
    * Chiliz
* Distributed File-Sharing/Streaming
    * Theta
    * Storj
    
The possibilities grow every year. New applications and utilities are being built to leverage the power of blockchain technology!

Further research (aka a *much* better explanation of blockchain) can be found here:
https://www.youtube.com/watch?v=bBC-nXj3Ng4